src/Framework/Framework/Resources/Scripts/tests/stateManagement.test.ts

```diff
@@ -692,3 +692,106 @@ test("state is frozen", () => {
+        if (typeof item().Id() != "number")
+            throw `Element ${label} does not have number Id`
+    }
+
+    const eventLog: any[] = []
+    const diffLog: any[] = []
+    const sub1 = arr.subscribe(newValue => {
+        eventLog.push(ko.toJS(newValue))
+        
+        newValue.forEach(validateElement)
+    });
+    const sub2 = arr.subscribe(diff => {
+        diffLog.push(ko.toJS(diff))
+        
+        diff.forEach((e, i) => {
+            validateElement(e.value, `${e.index}:${e.status}`)
+        })
+
+    }, null, "arrayChange")
+
+
+    try {
+        callback(arr, initial, eventLog, diffLog)
+    } finally {
+        sub1.dispose()
+        sub2.dispose()
+    }
+}
+
+test("ko.observableArray - clones pushed plain observable", () => {
+    setupValidatingArraySubscriptions((arr: any, initial) => {
+        arr.push(ko.observable({ Id: ko.observable(1) }))
+
+        arr.replace(arr().at(-1)!, ko.observable({ Id: ko.observable(10) }))
+
+        arr.splice(0, 0, ko.observable({ Id: ko.observable(2) }))
+
+        expect(arr.state).toEqual([{ Id: 2, $type: "t2" }, ...initial, { Id: 10, $type: "t2" }])
+        expect(arr.state).toEqual(ko.toJS(arr))
+    })
+})
+
+test("ko.observableArray - clones pushed plain objects", () => {
+    setupValidatingArraySubscriptions((arr: any, initial) => {
+        arr.push({ Id: 1 })
+
+        arr.replace(arr().at(-1)!, { Id: 10 })
+
+        arr.splice(0, 0, { Id: 2 })
+
+        expect(arr.state).toEqual([{ Id: 2, $type: "t2" }, ...initial, { Id: 10, $type: "t2" }])
+        expect(arr.state).toEqual(ko.toJS(arr))
+    })
+})
+
+test("ko.observableArray - clones pushed mixed observables", () => {
+    setupValidatingArraySubscriptions((arr: any, initial) => {
+        arr.push(ko.observable({ Id: 1 }))
+
+        arr.replace(arr().at(-1)!, { Id: ko.observable(10) })
+
+        arr.splice(0, 0, ko.observable({ Id: 2 }))
+
+        expect(arr.state).toEqual([{ Id: 2, $type: "t2" }, ...initial, { Id: 10, $type: "t2" }])
+        expect(arr.state).toEqual(ko.toJS(arr))
+    })
+})
+test("ko.observableArray - clones pushed dotvvm object taken from elsewhere", () => {
+    setupValidatingArraySubscriptions((arr: any, initial) => {
+        arr.push({ Id: 1 })
+        arr.push(vm.Array().at(-1)())
+        arr().at(-1)().Id(2)
+
+        expect(arr.state).toEqual([...initial, { Id: 1, $type: "t2" }, { Id: 2, $type: "t2" }])
+        expect(arr.state).toEqual(ko.toJS(arr))
+    })
+})
+
+// TODO: does not work yet
+// test("ko.observableArray - clones pushed dotvvm observable taken from elsewhere", () => {
+//     setupValidatingArraySubscriptions((arr: any, initial) => {
+//         arr.push({ Id: 1 })
+//         arr.push(vm.Array().at(-1))
+//         arr().at(-1)().Id(2)
+
+//         expect(arr.state).toEqual([...initial, { Id: 1, $type: "t2" }, { Id: 2, $type: "t2" }])
+//         expect(arr.state).toEqual(ko.toJS(arr))
+//     })
+// })
```

Comment:
This case still doesn't work and will be fixed in a subsequent PR.

I experimented with possible solutions and the one I like the most is to freeze the ko.observableArray underlying array and clone it on every operation.
The implementation in knockout.js is easy, it will slightly simplify our array handling logic, but it is a breaking change:

Currently, it is possible to call `array().push(1)`. The behavior will most likely be very broken, as it does not go through the validator and does not notify subscribers. In some cases, it might however work correctly.

If we make the array frozen, it will throw an exception. That is definitely desired, but I'd prefer to not push this as a 4.3 patch

Reply:
My guess is that in 99% of cases, directly modifying the underlying array is a bug and something people do intentionally. I do not insist on having it in 4.3, but I expect it will uncover a few bugs in Business Pack.